File: 438 Find All Anagrams in a String/findAnagrams.java
public class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int[] freqS = new int[26];
        int[] freqP = new int[26];
        List<Integer> index = new ArrayList<Integer>();
        
        for(int i=0; i<p.length(); i++)
            freqP[p.charAt(i)-'a']++;
        
        int l = 0, r = -1;
        while(l < s.length()){
            if(r-l+1 <= p.length() && r+1 < s.length())
                freqS[s.charAt(++r)-'a']++;
            else
                freqS[s.charAt(l++)-'a']--;
            if(Arrays.equals(freqS,freqP))
                index.add(l);
        }
        return index;
    }
}

File: 438 Find All Anagrams in a String/findAnagrams.py
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        # Define the list of frequencies in s and p
        freqS = [0] * 26
        freqP = [0] * 26
        index = []
        
        for x in range(len(p)):
            freqP[ord(p[x])%97] += 1
        
        l = 0
        r = -1
        while l < len(s):
            if r-l+1 <= len(p) and r+1 < len(s):
                r += 1
                freqS[ord(s[r])%97] += 1 # Note that in python it cannot directly recognize char as its ASCII number, need ord()
            else:
                freqS[ord(s[l])%97] -= 1
                l += 1
            if freqS == freqP:
                index.append(l)
        return index
                